coreTable/OwidTable.ts

```diff
@@ -281,70 +281,77 @@ export class OwidTable extends CoreTable<OwidRow, OwidColumnDef> {
-        const newDef = {
-            ...col.def,
-            type: ColumnTypeNames.Percentage,
-            values: col.parsedValues.map((val, index) => {
+        const newDefs = replaceDef(this.defs, [
+            {
+                ...col.def,
+                type: ColumnTypeNames.Percentage,
+            },
+        ])
+        const newColumnStore: CoreColumnStore = {
+            ...this.columnStore,
+            [columnSlug]: this.columnStore[columnSlug].map((val, index) => {
                 const timeTotal = timeTotals.get(timeValues[index])
                 if (timeTotal === 0) return InvalidCellTypes.DivideByZeroError
                 return (100 * (val as number)) / timeTotal!
             }),
         }
-
-        return new OwidTable(
-            this.columnStore,
-            replaceDef(this.defs, [newDef]),
-            {
-                parent: this,
-                tableDescription: `Transformed ${columnSlug} column to be % contribution of each entity for that time`,
-                transformCategory: TransformType.UpdateColumnDefs,
-            }
+        return this.transform(
+            newColumnStore,
+            newDefs,
+            `Transformed ${columnSlug} column to be % contribution of each entity for that time`,
+            TransformType.UpdateColumnDefs
         )
     }
 
     // If you want to see how much each column contributed to the entity for that year, use this.
     // NB: Uses absolute value. So if one entity added 100, and another -100, they both would have contributed "50%" to that year.
     // Otherwise we'd have NaN.
     toPercentageFromEachColumnForEachEntityAndTime(columnSlugs: ColumnSlug[]) {
+        columnSlugs = columnSlugs.filter((slug) => this.has(slug))
+        if (!columnSlugs.length) return this
+
         const newDefs = this.defs.map((def) => {
             if (columnSlugs.includes(def.slug))
                 return { ...def, type: ColumnTypeNames.RelativePercentage }
             return def
         })
-        return new OwidTable(
-            this.rows.map((row) => {
-                const newRow = {
-                    ...row,
-                }
-                const total = sum(
-                    columnSlugs
-                        .map((slug) => row[slug])
-                        .filter(isValid)
-                        .map((val) => Math.abs(val))
-                )
-                columnSlugs.forEach((slug) => {
-                    const value =
-                        total === 0
-                            ? InvalidCellTypes.DivideByZeroError
-                            : row[slug]
-                    newRow[slug] = isValid(value)
-                        ? (100 * value) / total
-                        : value
-                })
-                return newRow
-            }),
+
+        const columnStore = this.columnStore
+        const columnStorePatch: CoreColumnStore = {}
+
+        const totals = new Array(this.numRows).fill(0).map((_, i) =>
```

Comment:
Is `new Array(this.numRows).fill(0)` the fastest way in JS to create a range? I've been using the lodash `range`) method but figured it's probably not the fastest way.

Reply:
Yeah I should've gone with range, I just thought it's quicker to use vanilla `new Array(length)` but that doesn't work if you wanna map over it.

Reply:
You can still map over the indexes.My hunch is `new Array` may be faster but have to check. We could define our own range if so.

Reply:
(obviously not a priority, just wondering aloud)

Reply:
https://stackoverflow.com/a/35086350